File: python/merge_sort.py
def mergesort(list1):
  if len(list1) > 1:
  
    # To divide List
    mid = len(list1)//2
    left_list = list1[:mid]
    right_list = list1[mid:]
    mergesort(left_list)
    mergesort(right_list)
  
    # To Merge Lists
    i,j,k = 0,0,0
    while i < len(left_list) and j < len(right_list):
      if left_list[i] < right_list[j]:
        list1[k] = left_list[i]
        i += 1
        k += 1
      else:
        list1[k] = right_list[j]
        j += 1
        k += 1
  
    # to check left over on left list
    while i < len(left_list):
      list1[k] = left_list[i]
      i += 1
      k += 1
  
    # to check left over on right list
    while j < len(right_list):
      list1[k] = right_list[j]
      j += 1
      k += 1

list1 = [9,6,3,2,7,1]
mergesort(list1)
print(list1)

File: cpp/mergesort.cpp
#include<bits/stdc++.h>
using namespace std;
void merge(vector<int>&v,int low,int mid,int high){
vector<int>temp;
  int left=low;
  int right=mid+1;
  while(left<=mid&&right<=high){
    if(v[left]<=v[right]){
    temp.push_back(v[left]);
    left++;
    }
    else{
    temp.push_back(v[right]);
    right++;
    }
  }
    while(left<=mid){
    temp.push_back(v[left]);
    left++;
    }
    while(right<=high){
    temp.push_back(v[right]);
    right++;
    }
    for(int i=low;i<=high;i++){
      v[i]=temp[i-low];
    }
}
void mergesort(vector<int>&v,int low,int high){
 if(low==high) return;
 int mid=(low+high)/2;
 mergesort(v,low,mid);
 mergesort(v,mid+1,high);
 merge(v,low,mid,high);
}
int main()
{
  vector<int>v{3,1,2,4,1,5,2,6,4};
  int low=0;
  int high=v.size()-1;
  mergesort(v,low,high); 
  for(int i=0;i<v.size();i++){
    cout<<v[i]<<" ";
  }
  return 0;
}